enforcer/enforcer.go

```diff
@@ -0,0 +1,289 @@
+//
+// This package defines the different enforcer APIs and provides interfaces to
+// convert any request into a PADME Resource so it can be evaluated against the
+// policies known to the enforcer.
+package enforcer
+
+import (
+	"log"
+	"time"
+
+	"github.com/padmeio/padme/enforcer/plugins"
+	"github.com/padmeio/padme/enforcer/store"
+	"github.com/padmeio/padme/enforcer/utils"
+	"github.com/padmeio/padme/policy"
+)
+
+// loaddedPlugin holds the information of a plugin that has been loaded by this enforcer,
+// such as the enabled flag, plugin-specific configuration that needs to be known by the
+// enforcer, etc.
+type loadedPlugin struct {
+	plugins.Plugin
+	enabled bool
+}
+
+// Enforcer is the main implementation of a PADME Enforcer.
+type Enforcer struct {
+
+	// Location configures the location where teh current enforcer operates.
+	// This field will be used when matching policies and only the policies that
+	// apply to this location will be considered.
+	Location *policy.Location
+
+	// Store configures the repository where the policies for this enforcer
+	// are stored.
+	Store store.PolicyRepository
+
+	// TODO nacx: How to implement persistence for plugins and controllers
+	// in case an enforcer is restarted? (for example after recovering from a crash)
+
+	// Handlers are the list of controllers known to this enforcer that
+	// are subscribed to policy events
+	Handlers map[string]PolicyEventHandler
+
+	// RegisteredPlugins are the list of plugins this enforcer will delegate to when
+	// checking policies for an incoming resource
+	RegisteredPlugins map[string]*loadedPlugin
+
+	// resource is the representation of this Enforcer as a PADME Policy resource.
+	// This will be used in the Request Level Answer API to match policies that
+	// target this enforcer
+	resource *policy.Resource
+}
+
+// NewEnforcer builds a new Enforcer object with the given policy repository
+func NewEnforcer(store store.PolicyRepository, location *policy.Location, credentials *policy.Credential) Enforcer {
+	localAddresses, err := utils.LocalAddresses()
+	if err != nil {
+		log.Fatalf("error reading local addresses: %v", err)
+	}
+
+	var resource *policy.Resource
+	resource, err = utils.AddressesToResource(localAddresses, credentials)
+	if err != nil {
+		log.Fatalf("error building enforcer resource: %v", err)
+	}
+
+	return Enforcer{
+		Location:          location,
+		Store:             store,
+		Handlers:          make(map[string]PolicyEventHandler),
+		RegisteredPlugins: make(map[string]*loadedPlugin),
+		resource:          resource,
+	}
+}
+
+// Implementation of the Request Level Answer API
+
+// Answer is the implementation of the Enforcer Answer API. It takes an access request for a
+// given resource and evaluates it against the existing policies.
+func (e *Enforcer) Answer(properties []*policy.Rule, credential *policy.Credential) bool {
+	var bundle *policy.PolicyBundle
+	if bundle = e.Fetch(); bundle == nil {
+		return false
```

Comment:
This is the lack of context upon failure I was referring to in the PR description. I didn't want to change the API interface that much but we'd probably make all API methods return an error in addition to the regular response.

Reply:
not returning why in this case is intentional.... during mainline operation it is not necessary imo due to the fact that the system is mostly remotely deployed.  Though it might become necessary when we start thinking about how accounting fits in. Where it is definitely necessary is where we preflight requests when users are testing policies.  But I think that should be a different version of the call.  The next question in that case however is: should we report the first negative result or all negative results possible.

Reply:
accounting might be necessary to support auditing and compliance type of use cases. How to provide that efficiently might need a little more thought especially when there is a lot of nesting involved. 
With a different version of APIs, we might have to think through when one API will be used over the other.